skynet.go

```diff
@@ -57,43 +126,176 @@ type (
+	}
+
+	respBody, err := parseResponseBody(resp)
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not parse response body")
+	}
+
+	var apiResponse CreateSkykeyResponse
+	err = json.Unmarshal(respBody.Bytes(), &apiResponse)
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not unmarshal response JSON")
+	}
+
+	return Skykey(apiResponse), nil
+}
+
+// GetSkykey returns the given skykey. One of either name or id must be provided
+// -- the one that is not provided should be left blank.
+func GetSkykey(name, id string, opts GetSkykeyOptions) (Skykey, error) {
+	body := &bytes.Buffer{}
+	url := makeURL(opts.PortalURL, opts.PortalGetSkykeyPath)
+	url = fmt.Sprintf("%s?name=%s&id=%s", url, name, id)
+
+	resp, err := executeRequest(opts.ConnectionOptions, "GET", url, body)
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not execute request")
+	}
+
+	respBody, err := parseResponseBody(resp)
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not parse response body")
+	}
+
+	var apiResponse GetSkykeyResponse
+	err = json.Unmarshal(respBody.Bytes(), &apiResponse)
+	if err != nil {
+		return Skykey{}, errors.AddContext(err, "could not unmarshal response JSON")
+	}
+
+	return Skykey(apiResponse), nil
+}
+
+// ListSkykeys returns a list of all skykeys.
+func ListSkykeys(opts ListSkykeysOptions) ([]Skykey, error) {
+	url := makeURL(opts.PortalURL, opts.PortalListSkykeysPath)
+
+	resp, err := executeRequest(opts.ConnectionOptions, "GET", url, &bytes.Buffer{})
+	if err != nil {
+		return nil, errors.AddContext(err, "could not execute request")
+	}
+
+	respBody, err := parseResponseBody(resp)
+	if err != nil {
+		return nil, errors.AddContext(err, "could not parse response body")
+	}
+
+	var apiResponse ListSkykeysResponse
+	err = json.Unmarshal(respBody.Bytes(), &apiResponse)
+	if err != nil {
+		return nil, errors.AddContext(err, "could not unmarshal response JSON")
+	}
+
+	return apiResponse.Skykeys, nil
+}
+
 // Upload uploads the given generic data.
 func Upload(uploadData UploadData, opts UploadOptions) (string, error) {
 	// prepare formdata
 	body := &bytes.Buffer{}
 	writer := multipart.NewWriter(body)
 
-	url := fmt.Sprintf("%s/%s", strings.TrimRight(opts.PortalURL, "/"), strings.TrimLeft(opts.PortalUploadPath, "/"))
+	url := makeURL(opts.PortalURL, opts.PortalUploadPath)
 
 	var fieldname string
+	var filename string
 	if len(uploadData) == 1 {
 		fieldname = opts.PortalFileFieldName
 	} else {
 		if opts.CustomDirname == "" {
 			return "", errors.New("CustomDirname must be set when uploading multiple files")
 		}
 		fieldname = opts.PortalDirectoryFileFieldName
-		url = fmt.Sprintf("%s?filename=%s", url, opts.CustomDirname)
+		filename = opts.CustomDirname
+	}
+	// Always send the filename even if it's empty. This lets us always pass
+	// more URL parameters using &.
+	url = fmt.Sprintf("%s?filename=%s", url, filename)
```

Comment:
@peterjanbrone a blank filename won't create an error will it?

Reply:
it looks like there are several error cases for the filename being blank. The SDK shouldn't validate anything right? It is all pass through code. So maybe it is the comment that is throwing me off.

Shouldn't we be adding all fields to the url, regardless of if they are blank, and let the Sia API handle all the errors?

Then maybe the comment should just be updated to say
```
// The filename is set first and handles including ? in the url string. All subsequent parameters should include an &
```
Or something like that? I think saying `even if it's empty` implies that we might not set other parameters if they are empty.

Reply:
We use `Values.get` to grab the filename which treats values that aren't present as blank strings:

```go
// Get gets the first value associated with the given key.
// If there are no values associated with the key, Get returns
// the empty string. To access multiple values, use the map
// directly.
func (v Values) Get(key string) string
```

Reply:
Didn't see your second comment.

> Shouldn't we be adding all fields to the url, regardless of if they are blank, and let the Sia API handle all the errors?

Sounds reasonable, I don't think we send any other fields yet but will keep this in mind.

Will update the comment.